webhooks-extension/endpoints/credentials.go

```diff
@@ -0,0 +1,261 @@
+}
+
+func (r Resource) getAllCredentials(request *restful.Request, response *restful.Response) {
+	targetNamespace := request.QueryParameter("namespace")
+	if targetNamespace != "" {
+		targetNamespace := r.Defaults.Namespace
+		if targetNamespace == "" {
+			targetNamespace = "default"
+		}
+	}
+	if !r.verifyNamespaceExists(targetNamespace, response) {
+		return
+	}
+
+	// Get secrets from the resource K8sClient
+	secrets, err := r.K8sClient.CoreV1().Secrets(targetNamespace).List(metav1.ListOptions{})
+
+	if err != nil {
+		errorMessage := fmt.Sprintf("Error getting secrets from K8sClient: %s.", err.Error())
+		response.WriteErrorString(http.StatusInternalServerError, errorMessage)
+		logging.Log.Error(errorMessage)
+		return
+	}
+
+	// Parse K8s secrets to credentials
+	creds := []credential{}
+	emptyCred := credential{}
+	for _, secret := range secrets.Items {
+		cred := secretToCredential(&secret, true)
+		if cred != emptyCred {
+			creds = append(creds, cred)
+			logging.Log.Infof("getAllCredentials Found credential %+v\n", cred)
+		}
+	}
+
+	logging.Log.Infof("getAllCredentials returning +%v", creds)
+
+	// Write the response
+	response.AddHeader("Content-Type", "application/json")
+	response.WriteEntity(creds)
+}
+
+// Convert credential struct into K8s secret struct
+func credentialToSecret(cred credential, namespace string, response *restful.Response) *corev1.Secret {
+	// Create new secret struct
+	secret := corev1.Secret{}
+	secret.Type = corev1.SecretTypeOpaque
+	secret.SetNamespace(namespace)
+	secret.SetName(cred.Name)
+	secret.Data = make(map[string][]byte)
+	secret.Data["accessToken"] = []byte(cred.AccessToken)
+	if cred.SecretToken != "" {
+		secret.Data["secretToken"] = []byte(cred.SecretToken)
+	} else {
+		secret.Data["secretToken"] = getRandomSecretToken()
+	}
+	return &secret
+}
+
+var (
+	src = rand.NewSource(time.Now().UnixNano())
+)
+
+const (
+	tokenBytes   = "123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	tokenIdxBits = 6                   // 6 bits = 2^6 = 64 characters in tokenBytes
+	tokenIdxMask = 1<<tokenIdxBits - 1 // All 1-bits, as many as tokenIdxBits
+)
+
+// Generate a random 20-character string, returned as []byte.
+// With thanks to https://medium.com/@kpbird/golang-generate-fixed-size-random-string-dd6dbd5e63c0
+func getRandomSecretToken() []byte {
+	b := make([]byte, 20)
+	for i := 0; i < 20; {
+		idx := int(src.Int63() & tokenIdxMask)
+		if idx < len(tokenBytes) {
+			b[i] = tokenBytes[idx]
+			i++
+		}
+	}
+	return b
+}
+
+// Convert K8s secret struct into credential struct
+func secretToCredential(secret *corev1.Secret, mask bool) credential {
+	var cred credential
+	if secret.Data["accessToken"] != nil {
+		cred = credential{
+			Name:            secret.GetName(),
+			AccessToken:     string(secret.Data["accessToken"]),
+			SecretToken:     string(secret.Data["secretToken"]),
+			ResourceVersion: secret.GetResourceVersion(),
+		}
+		if mask {
+			cred.AccessToken = "********"
+		}
```

Comment:
would we also want to mask the secret token??

Reply:
I thought we might want to display it, particularly in the case where we generated it, in case it's needed for debugging. I also wondered about not hiding the first few characters of each accessToken to help users know which was which. It's a shame it's called 'secretToken' by the knative GitHub event source and not 'handshake token' : )